### ISSUE TITLE
"lint:ignore" comments are ignored by staticcheck

### ISSUE THREAD

**Issue by simonpasquier** [golangci/golangci-lint#741]
It seems that golangci-lint v1.19.0 ignores `//lint:ignore` comments. Tested with this very simple program:

```
package main

import (
        "fmt"
        "time"
)

func main() {
        //lint:ignore SA1004 ignore this!
        time.Sleep(1)
        fmt.Println("hello")
}
```

golangci-lint v1.19.0 fails with:
```
main.go:10:13: SA1004: sleeping for 1 nanoseconds is probably a bug. Be explicit if it isn't: time.Sleep(time.Nanosecond) (staticcheck)
        time.Sleep(1)
```

golangci-lint v1.18.0 succeeds as well as staticcheck 2019.2.3.

### System information

1. Version of golangci-lint: `golangci-lint has version 1.19.0 built from 27ac4c7 on 2019-09-23T20:40:17Z`
2. Config file: N/A
3. Go environment:
```
go version go1.13 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/simon/.cache/go-build"
GOENV="/home/simon/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/simon/workspace/"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/simon/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/simon/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/tmp/foo/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build335086001=/tmp/go-build -gno-record-gcc-switches"
```
4. Verbose output of running: 

```
INFO [config_reader] Config search paths: [./ /tmp/foo /tmp /] 
INFO [lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck] 
INFO [lintersdb] Optimized sublinters [staticcheck gosimple unused] into metalinter megacheck 
INFO [loader] Go packages loading at mode 1023 (exports_file|files|imports|name|types|types_info|compiled_files|deps|syntax|types_sizes) took 724.109724ms 
INFO [runner] worker.8 took 21.127µs with stages: typecheck: 17.201µs 
INFO [runner] worker.3 took 94.53µs with stages: varcheck: 81.118µs, structcheck: 8.362µs 
INFO [runner] worker.5 took 40.529µs with stages: deadcode: 33.001µs 
INFO [runner] worker.7 took 148.077µs with stages: ineffassign: 145.627µs 
INFO [runner] worker.2 took 1.143µs               
INFO [runner] worker.4 took 129.961µs with stages: errcheck: 111.743µs 
INFO [runner] worker.6 took 25.834526ms with stages: govet: 25.829686ms 
INFO [runner] worker.1 took 44.510349ms with stages: megacheck: 44.502055ms 
INFO [runner] Workers idle times: #2: 44.358168ms, #3: 44.380638ms, #4: 44.284389ms, #5: 44.373782ms, #6: 18.681965ms, #7: 44.369167ms, #8: 44.390642ms 
INFO [runner] Processors filtering stat (out/in): filename_unadjuster: 1/1, autogenerated_exclude: 1/1, exclude-rules: 1/1, path_shortener: 1/1, cgo: 1/1, source_code: 1/1, path_prettifier: 1/1, uniq_by_line: 1/1, max_per_file_from_linter: 1/1, max_same_issues: 1/1, skip_dirs: 1/1, identifier_marker: 1/1, exclude: 1/1, nolint: 1/1, diff: 1/1, max_from_linter: 1/1, skip_files: 1/1 
INFO [runner] processing took 318.825µs with stages: exclude: 134.356µs, identifier_marker: 57.768µs, autogenerated_exclude: 42.228µs, source_code: 32.72µs, skip_dirs: 17.891µs, nolint: 10.125µs, cgo: 5.393µs, path_prettifier: 4.858µs, uniq_by_line: 3.079µs, filename_unadjuster: 2.773µs, max_same_issues: 2.363µs, max_from_linter: 1.579µs, path_shortener: 1.323µs, skip_files: 740ns, max_per_file_from_linter: 640ns, diff: 619ns, exclude-rules: 370ns 
main.go:10:13: SA1004: sleeping for 1 nanoseconds is probably a bug. Be explicit if it isn't: time.Sleep(time.Nanosecond) (staticcheck)
        time.Sleep(1)
                   ^
INFO File cache stats: 1 entries of total size 129B 
INFO Memory: 9 samples, avg is 121.8MB, max is 136.7MB 
INFO Execution took 774.097404ms                  
```

**Comment by bmhatfield** [comment]
Couple notes here, because I just hit this:

1. It's actually intermittent. Locally and in CircleCI this hits about 75% of the time
2. `// nolint:staticcheck` seems to work as a workaround.

**Comment by denisvmedia** [comment]
I also faced this problem. Well, a better workaround would be disabling `staticcheck` at all and running the updated version from the [original source](https://github.com/dominikh/go-tools).

**Comment by jirfag** [comment]
I guess we incorrectly configure `staticcheck`. Contributions [are welcome](https://golangci-lint.run/contributing/quick-start/).

**Comment by ahmedtalhakhan** [comment]
Any progress here?

**Comment by jsvisa** [comment]
Seems it's introduced by https://github.com/golangci/golangci-lint/pull/699

**Comment by simonpasquier** [comment]
From https://golangci-lint.run/usage/false-positives/ + some experiments, you have to do this now:

```go
package main

import (
        "fmt"
        "time"
)

func main() {
       //nolint:staticcheck // SA1004 ignore this!
        time.Sleep(1)
        fmt.Println("hello")
}
```

**Comment by jsvisa** [comment]
> //nolint:staticcheck // SA1004 ignore this!

@simonpasquier This would be disabling staticcheck at all, not only the SA1004 of staticcheck.

**Comment by uhthomas** [comment]
@simonpasquier can this be re-opened? I don't believe using `//nolint:staticcheck // SA1004 ignore this!` resolves this issue as it disables all static checks.

**Comment by bconway** [comment]
I would also like to see this revisited. I have the following code (or similar) in a few places due to imported protobuf definitions that use the older Go protobuf API version:

```
//lint:ignore SA1019 // third-party dependency
"github.com/golang/protobuf/proto"
```

I also like to use both staticcheck standalone and golangci-lint, and I don't believe disabling staticcheck entirely for the file is a great workaround. Has there been any determination on why these directives are ignored when staticcheck is called from within golangci-lint?

**Comment by NYARAS** [comment]
This should work

```golang
package main

import (
        "fmt"
        "time"
)

func main() {
      //lint:ignore SA1004 trying out line-based linter directives
        time.Sleep(1)
        fmt.Println("hello")
}
```

**Comment by shakefu** [comment]
> This should work
> 
Against what version? Definitely not working for me.

**Comment by mikeatlas** [comment]
Currently have `staticcheck 2020.2.3 (v0.1.3)` in my path, but in VSCode:

![image](https://user-images.githubusercontent.com/12704/134738622-e22c7366-6f39-4ca5-a96e-14de953992ef.png)

Ignoring `S1034` for me is not working.

**Comment by uloureiro** [comment]
I couldn't get it to work out of the box as well. My need was to ignore a certain directive just in certain files as a whole and neither using `//nolint:staticcheck` (as this disables everything as others already experienced) nor `//lint:file-ignore SA1019 disable warnings for deprecated code` (which does not even seem to work) is an option.

The only way I could work around it was by adding the exclude rule I wanted in the `.golangci.yaml` file in my project.

Here is an example that worked to ignore a single directive in a certain file:

```yaml
issues:
  exclude-rules:
    - path: path/to/your/file/your_file.go
      linters:
        - staticcheck
      text: SA1019
```

**Comment by rulisastra** [comment]
> I couldn't get it to work out of the box as well. My need was to ignore a certain directive just in certain files as a whole and neither using `//nolint:staticcheck` (as this disables everything as others already experienced) nor `//lint:file-ignore SA1019 disable warnings for deprecated code` (which does not even seem to work) is an option.
> 
> The only way I could work around it was by adding the exclude rule I wanted in the `.golangci.yaml` file in my project.
> 
> Here is an example that worked to ignore a single directive in a certain file:
> 
> ```yaml
> issues:
>   exclude-rules:
>     - path: path/to/your/file/your_file.go
>       linters:
>         - staticcheck
>       text: SA1019
> ```

please help me...
how do you specify the path if the project is a team project? if maybe... input it as a variable or smth?

**Comment by aarongable** [comment]
This bug is not fixed; can it please be re-opened?

When I run staticcheck standalone against my code, it correctly ignores lines annotated with `//lint:ignore ID reason`. When I run golangci-lint with staticcheck enabled, it reports those same lines as failures.

The workarounds posted above are incomplete: using [the one posted by uloureiro](https://github.com/golangci/golangci-lint/issues/741#issuecomment-997974984) disables the given check for the entire file, not just the single line; [the one posted simonpasquier](https://github.com/golangci/golangci-lint/issues/741#issuecomment-775267807) disables all staticcheck checks for that line, not just the single check. The [example posted by NYARAS](https://github.com/golangci/golangci-lint/issues/741#issuecomment-893139041) **should** work, but it does not: that's the whole point of this issue.

Please re-open and investigate why golangci-lint causes staticcheck to not respect its own `//lint:ignore` directives.

**Comment by bconway** [comment]
Same, still fails here.

**Comment by ldez** [comment]
I will try to provide a better overview.

`staticcheck` is a binary **and** a set of rules, those rules are grouped in "category": `staticcheck`, `gosimple`, `stylecheck` and `unused`.

- staticcheck: https://github.com/dominikh/go-tools/blob/331bcca40621580531c9b8a9cb7d5c2c06c709d2/staticcheck/analysis.go#L23
- stylecheck: https://github.com/dominikh/go-tools/blob/331bcca40621580531c9b8a9cb7d5c2c06c709d2/stylecheck/analysis.go#L14
- simple: https://github.com/dominikh/go-tools/blob/331bcca40621580531c9b8a9cb7d5c2c06c709d2/simple/analysis.go#L11
- unused: https://github.com/dominikh/go-tools/blob/331bcca40621580531c9b8a9cb7d5c2c06c709d2/unused/unused.go#L418

Inside golangci-lint we use the sets of rules, not the binary.
And all functions are not exported in "staticcheck".

The `lint` directive is something specific to "staticcheck the binary".

We recommend using the `nolint` directive that we provide inside golangci-lint.

https://golangci-lint.run/usage/false-positives/#nolint

**Comment by rulisastra** [comment]
after my work got delayed by this issue, I finally passed the Jenkins CI/CD build by creating a `.golangci.yml` file and add 
```
linters:
  enable:
    - misspell
    - typecheck
    - etc...
    
issues:
  exclude:
    - ^G104
  exclude-rules:
    - linters:
        - staticcheck
      text: "SA1019:"
```
with SA1019 comment on top of the deprecated method is used.
```
//lint:ignore SA1019 Ignore the deprecation warnings
	ok, err := yourDeprecatedMethod(input)
```

**Comment by Integralist** [comment]
👋🏻 3.5 years later and it looks like this issue isn't going to be reopened :-/

None of the suggestions presented in this issue actually work, here's a screenshot of the last one I tried...

<img width="730" alt="Screenshot 2023-03-28 at 10 37 48" src="https://user-images.githubusercontent.com/180050/228195221-4ec37a69-6d3a-45fc-84ee-af6668ac5b77.png">

**Comment by denisvmedia** [comment]
> 👋🏻 3.5 years later and it looks like this issue isn't going to be reopened :-/
> 
> None of the suggestions presented in this issue actually work, here's a screenshot of the last one I tried...
> 
> <img alt="Screenshot 2023-03-28 at 10 37 48" width="730" src="https://user-images.githubusercontent.com/180050/228195221-4ec37a69-6d3a-45fc-84ee-af6668ac5b77.png">

I bet you didn't try my solution: https://github.com/golangci/golangci-lint/issues/741#issuecomment-596236953.
If it doesn't work for you, you then need to report it to the original linter author.

**Comment by Integralist** [comment]
Hi @denisvmedia 

I did see your comment but I didn't quite understand it to be honest.

So to clarify, your suggestion is to avoid using staticcheck via golangci-lint and to just use it directly, is that right?

If so then I guess that is possible to do as it looks like null-ls (which I also use) has a builtin for staticcheck https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/lua/null-ls/builtins/diagnostics/staticcheck.lua

**Comment by lifenjoiner** [comment]
Here is the solution to eliminate the error itself on `time.Sleep(1)`, rather than making `//lint:ignore <CODE> <REASON>` work:
```
time.Sleep(time.Nanosecond)
```

Just a quick solution, but not mentioned above.

**Comment by sfate** [comment]
## Solution for deprecated packages without `//nolint:staticcheck`

I'll leave this for future reference to anyone who's dealing (_might be dealing_) with deprecated package issues.


**TL;DR**
Add `exclude-rules` with specified `path` of file and linter issue`text` (full offense message) to `.golangci.yml`.

### Details

Previously we were dealing with deprecated `crypto/openpgp` package like that:
```go
//nolint:staticcheck
import "golang.org/x/crypto/openpgp"
```

Which does not work anymore for `golangci-lint` and it provides no clear details:
```bash
path/to/file.go:3: directive `//nolint:staticcheck` is unused for linter "staticcheck" (nolintlint)
                //nolint:staticcheck
                ^
make: *** [golangci-lintint] Error 1
```

Removing `//nolint:staticcheck` directive brings up origins of the issue with deprecated package:
```bash
path/to/file.go:3:2: SA1019: "golang.org/x/crypto/openpgp" is deprecated: this package is unmaintained except for security fixes. New applications should consider a more focused, modern alternative to OpenPGP for their specific task. If you are required to interoperate with OpenPGP systems and need a maintained package, consider a community fork. See https://golang.org/issue/44226. (staticcheck)
        "golang.org/x/crypto/openpgp"
        ^
make: *** [golangci-lintint] Error 1
```

### Solution

The only viable solution to achieve disabling this check is to add `exclude-rules` to `.golangci.yml`:
(_I strongly suggest to use full check text to avoid unexpected ignores further_)
```yaml
issues:
  exclude-rules:
    - path: path/to/file.go
      text: 'SA1019: "golang.org/x/crypto/openpgp" is deprecated: this package is unmaintained except for security fixes. New applications should consider a more focused, modern alternative to OpenPGP for their specific task. If you are required to interoperate with OpenPGP systems and need a maintained package, consider a community fork. See https://golang.org/issue/44226.'
```

**Comment by gluax** [comment]
I can't believe this issue is still marked as closed despite not having a way to inline comment ignore a specific line. I see editing the yaml file as a non-solution to this problem.

Is this project dead or something?

**Comment by ldez** [comment]
I already provided all the information here: https://github.com/golangci/golangci-lint/issues/741#issuecomment-1016936027

This issue was about `lint` directives, those directives are not supported by golangci-lint because they are something specific to "staticcheck the binary".

Golangci-lint provides `nolint` directives: https://golangci-lint.run/usage/false-positives/#nolint-directive

---

This issue was closed by the original author because he found a solution (the usage of `nolint` directives).

---

About the deprecated packages:

</details>

<details><summary>Without `//nolint:staticcheck`</summary>

```go
package main

import (
	"fmt"

	"golang.org/x/crypto/openpgp"
)

func main() {
	fmt.Println(openpgp.SignatureType)
}
```

```console
$ golangci-lint version 
golangci-lint has version 1.55.2 built with go1.21.3 from e3c2265f on 2023-11-03T12:59:25Z
$ golangci-lint run --no-config --enable-all=false --enable=staticcheck
main.go:6:2: SA1019: "golang.org/x/crypto/openpgp" is deprecated: this package is unmaintained except for security fixes. New applications should consider a more focused, modern alternative to OpenPGP for their specific task. If you are required to interoperate with OpenPGP systems and need a maintained package, consider a community fork. See https://golang.org/issue/44226. (staticcheck)
        "golang.org/x/crypto/openpgp"
        ^
```

</details>


<details><summary>Without `//nolint:staticcheck`</summary>

```go
package main

//nolint:staticcheck
import (
	"fmt"

	"golang.org/x/crypto/openpgp"
)

func main() {
	fmt.Println(openpgp.SignatureType)
}
```

```console
$ golangci-lint version 
golangci-lint has version 1.55.2 built with go1.21.3 from e3c2265f on 2023-11-03T12:59:25Z
$ golangci-lint run --no-config --enable-all=false --enable=staticcheck
$
```

</details>

As you can see the directive `nolint` works to handle this problem.

**Comment by gluax** [comment]
@ldez ah my bad, that linked comment didn't make it clear to me that the comment functionality was fixed. It seemed as if it was only fixed from editing a yaml file.

**Comment by fademan01** [comment]
> I bet you didn't try my solution: [#741 (comment)](https://github.com/golangci/golangci-lint/issues/741#issuecomment-596236953). If it doesn't work for you, you then need to report it to the original linter author.
It worked for me with your solution. just run `go install honnef.co/go/tools/cmd/staticcheck@2022.1` then add `//nolint:staticcheck  ` to top of function then the error message disappeared. Thanks

**Comment by alexbozhenko** [comment]
@ldez 
re https://github.com/golangci/golangci-lint/issues/741#issuecomment-1016936027

Maybe [some people think](https://github.com/search?q=%2F%28%3F-i%29%5C%2F%5C%2Fnolint%3Astaticcheck+%5C%2F%5C%2F+S%5BAT%5D%5B0-9%5D%2F+&type=code) that adding statick check rule name works the same way as for `//lint:ignore` directive...

Would you be open to a PR documenting this(basically, explanation you gave in the linked comment) in the false-positives doc?

**Comment by ldez** [comment]
`//nolint:staticcheck // SA1019 foo foo` is a perfectly right usage.

The comment after the directive explains the reason for the usage `nolint` directive, so having the name of the staticcheck's rule is a good idea.

IMO, this syntax is not related to `lint:ignore` syntax.

---

As GitHub collapsed and so hidden my previous answers I will add links here:
- https://github.com/golangci/golangci-lint/issues/741#issuecomment-1016936027
- https://github.com/golangci/golangci-lint/issues/741#issuecomment-1901329406

**Comment by MisguidedEmails** [comment]
> `//nolint:staticcheck // SA1019 explanation` is a perfectly right usage.

I would disagree as:
- `staticcheck` will raise a linter error if the ignored lint is not actually valid anymore. This will not as it disables `staticcheck` entirely.
- If there is another issue, or a different issue, on that same line `staticcheck` will not detect it as it's disabled.

I don't think disabling the entire linter for a line is a suitable solution to this problem.

Does `golangci-lint` have no intention in supporting the `nolint` for specific error messages in specific linters, or supporting the native "no lint" functionality that linters have?

**Comment by ldez** [comment]
My previous comment was about the suggestion to provide more details inside a comment of a `nolint` directive.
This was not something where you can agree or disagree; this is not about the `nolint` directive behavior.

This issue is neither about `nolint` directives nor if it's "right or wrong" to not handle those directives per internal metalinter (i.e., staticcheck, revive, etc.) rule.

We have a dedicated issue about the topic of a new directive syntax.
https://github.com/golangci/golangci-lint/issues/1658

I locked this issue because of off-topic comments.
